boa/vm/py_evm.py

```diff
@@ -0,0 +1,579 @@
+
+        # we usually want to reset the trace data structures
+        # but sometimes don't, give caller the option.
+        if reset_traces:
+            self.sha3_trace = {}
+            self.sstore_trace = {}
+
+        # patch in tracing opcodes
+        c.opcodes[0x20] = Sha3PreimageTracer(c.opcodes[0x20], self)
+        c.opcodes[0x55] = SstoreTracer(c.opcodes[0x55], self)
+
+    def _trace_sha3_preimage(self, preimage, image):
+        self.sha3_trace[image] = preimage
+
+    def _trace_sstore(self, account, slot):
+        self.sstore_trace.setdefault(account, set())
+        # we don't want to deal with snapshots/commits/reverts, so just
+        # register that the slot was touched and downstream can filter
+        # zero entries.
+        self.sstore_trace[account].add(slot)
+
+    def enable_fast_mode(self, flag: bool = True):
+        if flag:
+            patch_pyevm_state_object(self.vm.state)
+        else:
+            unpatch_pyevm_state_object(self.vm.state)
+
+    def fork_rpc(
+        self,
+        rpc: RPC,
+        reset_traces: bool,
+        fast_mode_enabled: bool,
+        block_identifier: str,
+        **kwargs,
+    ):
+        account_db_class = AccountDBFork.class_from_rpc(rpc, block_identifier, **kwargs)
+        self._init_vm(self.env, account_db_class, reset_traces, fast_mode_enabled)
+        block_info = self.vm.state._account_db._block_info
+
+        self.vm.patch.timestamp = int(block_info["timestamp"], 16)
+        self.vm.patch.block_number = int(block_info["number"], 16)
+        # TODO patch the other stuff
+
+        self.vm.state._account_db._rpc._init_mem_db()
+
+    @property
+    def is_forked(self):
+        return self.vm.__class__._state_class.account_db_class == AccountDBFork
+
+    def _set_account_db_class(self, account_db_class: type):
+        self.vm.__class__._state_class.account_db_class = account_db_class
+
+    def get_gas_meter_class(self):
+        return self.vm.state.computation_class._gas_meter_class
+
+    def set_gas_meter_class(self, cls: type):
+        self.vm.state.computation_class._gas_meter_class = cls
+
+    def get_balance(self, address: Address):
+        return self.vm.state.get_balance(address.canonical_address)
+
+    def set_balance(self, address: Address, value):
+        self.vm.state.set_balance(address.canonical_address, value)
+
+    def get_code(self, address: Address) -> bytes:
+        return self.vm.state.get_code(address.canonical_address)
+
+    def get_gas_limit(self):
+        return self.vm.state.gas_limit
+
+    # advanced: reset warm/cold counters for addresses and storage
+    def reset_access_counters(self):
+        self.vm.state._account_db._reset_access_counters()
+
+    def snapshot(self) -> Tuple[Hash32, JournalDBCheckpoint]:
+        return self.vm.state.snapshot()
+
+    def anchor(self):
+        return self.vm.patch.anchor()
+
+    def revert(self, snapshot_id: Tuple[Hash32, JournalDBCheckpoint]) -> None:
```

Comment:
i'd prefer the snapshot_id to be annotated as `Any`, to hide the implementation details